src/core/simulation.jl

```diff
@@ -688,3 +691,145 @@ function execute!(sim::Simulation; kwargs...)
 end
+
+struct SimulationSerializationWrapper
+    steps::Int
+    stages::Dict{String, StageSerializationWrapper}
+    initial_time::Union{Nothing, Dates.DateTime}
+    sequence::Union{Nothing, SimulationSequence}
+    simulation_folder::String
+    name::String
+end
+
+"""
+    serialize(simulation::Simulation, path = ".")
+
+Serialize the simulation to a directory in path.
+
+Return the serialized simulation directory name that is created.
+
+# Arguments
+- `simulation::Simulation`: simulation to serialize
+- `path = "."`: path in which to create the serialzed directory
+- `force = false`: If true, delete the directory if it already exists. Otherwise, it will
+   throw an exception.
+"""
+function serialize(simulation::Simulation; path = ".", force = false)
+    directory = joinpath(path, "simulation-$(simulation.name)")
+    stages = Dict{String, StageSerializationWrapper}()
+
+    orig = pwd()
+    if isdir(directory) || ispath(directory) && !force
+        throw(ArgumentError("$directory already exists. Please delete it or pass force = true."))
+    end
+    rm(directory, recursive = true, force = true)
+    mkdir(directory)
+    cd(directory)
+
+    try
+        for (key, stage) in simulation.stages
+            if isnothing(stage.internal)
+                throw(ArgumentError("stage $(stage.number) has not been built"))
+            end
+            sys_filename = "system-$(IS.get_uuid(stage.sys)).json"
+            # Skip serialization if multiple stages have the same system.
+            if !ispath(sys_filename)
+                PSY.to_json(stage.sys, sys_filename)
+            end
+            stages[key] = StageSerializationWrapper(
+                stage.template,
+                sys_filename,
+                stage.internal.psi_container.settings_copy,
+                typeof(stage),
+            )
+        end
+    finally
+        cd(orig)
+    end
+
+    filename = joinpath(directory, SIMULATION_SERIALIZATION_FILENAME)
+    obj = SimulationSerializationWrapper(
+        simulation.steps,
+        stages,
+        simulation.initial_time,
+        simulation.sequence,
+        simulation.simulation_folder,
+        simulation.name,
+    )
+    Serialization.serialize(filename, obj)
+    @info "Serialized simulation" simulation.name directory
+    return directory
+end
+
+"""
+    Simulation(directory::AbstractString; kwargs...)
+
+Constructs Simulation from a serialized directory. Callers should pass any kwargs here that
+they passed to the original Simulation.
+
+# Arguments
+- `directory::AbstractString`: the directory returned from the call to serialize
+- `stage_info::Dict`: Two-level dictionary containing stage parameters that cannot be
+  serialized. The outer dict should be keyed by the stage name. The inner dict must contain
+  'optimizer' and may contain 'jump_model'. These should be the same values used for the
+  original simulation.
+"""
+function Simulation(directory::AbstractString, stage_info::Dict; kwargs...)
```

Comment:
Accepting kwargs here is not idea.  I suggest we make a change to the Simulation constructor.
- We could make initial_times an optional parameter to it, defaulting to nothing.
- Remove kwargs from Simulation().
- Remove kwargs this constructor.  initial_times will have been serialized, so we will get it back. There won't be any ambiguity.

Reply:
Ok, we can address this after merging. I will make initial_time an optional parameter.

Reply:
Fixed it here.